internal/api/impl/v1/dashboard/schemas/validator.go

```diff
@@ -127,70 +117,121 @@ func (v *validator) Validate(panels map[string]json.RawMessage) error {
 
-		// build Value from the Instance
-		schema := v.context.BuildInstance(buildInstance)
-		if schema.Err() != nil {
-			logrus.WithError(schema.Err()).Errorf("Error during build for %s, skipping this schema", schemaPath)
+		// build the CUE Value from the files
+		schema, err := buildChartValueFromFiles(v.context, schemaFiles)
+		if err != nil {
+			logrus.WithError(err).Errorf("Not able to build CUE Value for %s, skipping this chart", schemaPath)
 			continue
 		}
 
-		// check if the schema fulfils the base panel requirements
-		unified := v.baseDef.Unify(schema)
-		if unified.Err() != nil {
-			logrus.WithError(unified.Err()).Errorf("Error during schema validation for %s, skipping this schema", schemaPath)
+		// check if Kind is defined + if another schema for the same Kind was already registered
+		kind, _ := schema.LookupPath(cue.ParsePath(kindCuePath)).String()
+		if len(kind) == 0 {
+			logrus.Errorf("Expected %s property missing for %s, skipping this chart", kindCuePath, schemaPath)
+			logrus.Tracef("%+v", schema)
 			continue
 		}
-
-		// check if another schema for the same Kind was already registered
-		kind, _ := schema.LookupPath(cue.ParsePath("kind")).String()
 		if _, ok := v.schemas.Load(kind); ok {
-			logrus.Errorf("Conflict caused by %s: a schema already exists for kind %s, skipping this schema", schemaPath, kind)
+			logrus.Errorf("Conflict caused by %s: a schema already exists for kind %s, skipping this chart", schemaPath, kind)
+			logrus.Tracef("%+v", schema)
+			continue
+		}
+
+		// at this stage everything is fine, so we add the "generator" CUE file & rebuild the CUE Value, in order to
+		// generate the final panel def (i.e a disjunction with 1 alternative for each kind of datasource+query)
+		schemaFiles = append(schemaFiles, filepath.Join(v.schemasConf.Path, generatorFile))
+		schema, err = buildChartValueFromFiles(v.context, schemaFiles)
+		if err != nil {
+			logrus.WithError(err).Errorf("Not able to use the generator for %s, skipping this chart", schemaPath)
 			continue
 		}
 
 		v.schemas.Store(kind, schema)
-		logrus.Debugf("Loaded schema %s from file %s: %+v", kind, schemaPath, schema)
+		logrus.Debugf("Loaded schema %s from dir %s: %+v", kind, schemaPath, schema)
 	}
 
 	logrus.Info("Schemas list (re)loaded")
 }
+
+// listCueFiles retrieves the cue files available at the given location and return a list containing their paths
+func listCueFiles(path string) ([]string, error) {
+	var cueFiles []string
+	err := filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+		if filepath.Ext(path) == ".cue" {
+			cueFiles = append(cueFiles, path)
+			logrus.Tracef("%s registered in CUE files list", path)
+		}
+		return nil
+	})
+	return cueFiles, err
+}
+
+// buildChartValueFromFiles builds a CUE Value representing a chart, from the provided list of CUE files
+func buildChartValueFromFiles(context *cue.Context, schemaFiles []string) (cue.Value, error) {
```

Comment:
Well not really, this would just remove the context from the signature, I'm calling this method 2 times and the first time is just to pass the checks on the schema, it is not stored in the map. Thus I prefer to pass around the context as a param, rather than to attach the func to the validator struct for which I have no use of the other properties (schemas map & schemasConf)